core/src/commonMain/kotlin/com/powersync/db/internal/InternalDatabaseImpl.kt

```diff
@@ -189,21 +160,31 @@ internal class InternalDatabaseImpl(
     override fun <RowType : Any> watch(
         sql: String,
         parameters: List<Any?>?,
+        throttleMs: Long?,
         mapper: (SqlCursor) -> RowType,
-    ): Flow<List<RowType>> {
-        val tables =
-            getSourceTables(sql, parameters)
-                .map { toFriendlyTableName(it) }
-                .filter { it.isNotBlank() }
-                .toSet()
-        return watchQuery(
-            query = sql,
-            parameters = parameters?.size ?: 0,
-            binders = getBindersFromParams(parameters),
-            mapper = mapper,
-            tables = tables,
-        ).asFlow().mapToList(scope.coroutineContext)
-    }
+    ): Flow<List<RowType>> =
+        // Use a channel flow here since we throttle (buffer used under the hood)
+        // This causes some emissions to be from different scopes.
+        channelFlow {
+            // Fetch the tables asynchronously with getAll
+            val tables =
+                getSourceTables(sql, parameters)
+                    .filter { it.isNotBlank() }
+                    .toSet()
+
+            updatesOnTables()
+                // onSubscription here is very important.
+                // This ensures that the initial result and all updates are emitted.
```

Comment:
Is this any different to just moving the `send` out like this? The `channelFlow` is cold too, so it would run for each subscriber.

```Kotlin
send(getAll(sql, parameters = parameters, mapper = mapper))
updatesOnTables()
  .filter { ... }
  .throttle()
  .collect { sendAll(...) }
```

I think the important bit is that we need to schedule another emission if the initial query and an update on a watched table is happening concurrently.
But `onSubscription` doesn't appear to do that, looking at its implementation it runs the callback first and then subsribes to the remaining flow.

Reply:
`onSubscribe` is actually fine as discussed offline.